frontend/src/metabase/query_builder/components/notebook/NotebookStep/NotebookStep.tsx

```diff
@@ -104,7 +105,7 @@ function NotebookStep({
   } = STEP_UI[step.type] || {};
 
   const color = getColor();
-  const canPreview = Boolean(step.getPreviewQuery);
+  const canPreview = Lib.canRun(step.query);
```

Comment:
`Boolean(step.getPreviewQuery)` and `step.active && step.visible` lead the same result. 

https://github.com/metabase/metabase/blob/master/frontend/src/metabase/query_builder/components/notebook/lib/steps.ts#L281-L286

`Lib.canRun` is to determine if the query can run, but what if step is not visible? I'd double check at least by running e2e tests

Reply:
you can check on `step.active` - if we render it, it's visible. `canRun` can return true even for not active step (e.g. expression step when there is no expression)

Reply:
I don't mind, all my comments were about the previous implementation and my initial comment was about the wrong semantics (we checked that there was a function passed and based on this we show or hide preview button). I like dropping old code, but sometimes there is a reason why the code works like this and not differently. 

so once again, if can_preview works, that's great

Reply:
> all my comments were about the previous implementation and my initial comment was about the wrong semantics

Oh, got it now. Thanks for the clarification.